### ISSUE TITLE
[Docs] Not really clear documentation

### ISSUE THREAD

**Issue by GitDruid** [microsoft/botframework-solutions#3579]
/overview/whats-new/1.0/enable-sso-with-skills-using-oauthcredentials-setting/

1) Is the approach suggested in this doc page compatible with the one used here?

https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-authentication-sso?view=azure-bot-service-4.0&tabs=csharp%2Ceml

2) Having the skill config knowing about the appId, appSecret and connection of the calling bot, does not prevent to have scenarios where I want to use skills developed by someone else?

**Comment by lauren-mills** [comment]
Looking into this. My first thought is that that are essentially the same approach but I'm confirming.

**Comment by GitDruid** [comment]
I don't think so, @lauren-mills .
[Here ](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-authentication-sso?view=azure-bot-service-4.0&tabs=csharp%2Ceml) skills and main bot have their own AD apps and oauth connections.

However I was able to have the [sample ](https://github.com/microsoft/BotBuilder-Samples/tree/master/experimental/sso-with-skills) working but still not able in the last two days to replicate the same behavior on a Virtual Assistant + Skills scenario. 
Not sure if due to codebase differences, azure settings or other.

In particular, in the VA I noticed that the token exchange does not even start without configuring this part in the appsettings.json:
```
"tokenExchangeConfig": {
  "connectionName": "",
  "provider": ""
},
```

that is injiected in the startup.cs:

```
// Configure TokenExchangeConfig for SSO
if (settings.TokenExchangeConfig != null)
{
  services.AddSingleton<ITokenExchangeConfig>(settings.TokenExchangeConfig);
}
```

Regarding this last setting I was not able to find any documentation. **Do you know if there is something about it?**

For instance, after a while I realized that "provider" should not be "Azure Active Directory v2" but "30dd229c-58e3-4a48-bdfd-91ec48eb906c" instead. Even with this setting, the token exchange process starts but rise an InternalServerError exception (see attachment).

In conclusion... quite frustrating days :-)

[Exception on token exchange.txt](https://github.com/microsoft/botframework-solutions/files/4967634/Exception.on.token.exchange.txt)

**Comment by lzc850612** [comment]
@GitDruid 

To your original question, the answer to both questions is the two things are different approaches that we support SSO. 

The approach under /overview/whats-new/1.0/enable-sso-with-skills-using-oauthcredentials-setting/ is to enable first party skills to easily share OAuth Connections (essentially trust boundary) with VA. This is going to be mostly used by enterprise scenarios where both VA and skills are owned by the same company/organization that makes it ok to share appid/password among them. This provide an easy way of reusing the same trust boundary for OAuth purpose, without having to get permission multiple times for skills Microsoft Apps and making sure skills are in the same trust boundary with VA. 

The approach talked about https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-authentication-sso?view=azure-bot-service-4.0&tabs=csharp%2Ceml is another approach which relies on token exchange mechanism. This is usually used in SSO through clients (webchat, Teams etc) to achieve SSO by passing in a token from the client and VA and/or skill uses that later to exchange for a token to use. This approach can support 3rd party skill where VA and skill are not in the same trust boundary but through token exchange mechanism can pass an exchangeable token over the wire for VA/skill to exchange. 

The reason the tokenExchangeConfig configuration setting section is not included in the latest VA is intentional because the SDK team is evaluating on including the TokenExchange capability into core SDK and logically it should. We want to include the capability in VA for some customers that want to taste this first without having it in SDK in advance, which is why we have the code to handle it, but we don't want everyone to use it before the SDK approach is put in which is why we deliberately hide the tokenExchangeConfig setting from appsettings. It's also why we didn't put any documentation because we'll have to change it later and cause confusion to our customers.

**Comment by GitDruid** [comment]
> @GitDruid
> 
> To your original question, the answer to both questions is the two things are different approaches that we support SSO.

Thank you for the answer Ted @lzc850612 , very appreciated.
However I made several tries without success. Mainly these are the two major issues I've encountered:

1) Magic code still appears (both with emulator and with web console, I've never had VA+Skill SSO scenario working without magic code being requested. While the solution descripted [here](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-authentication-sso?view=azure-bot-service-4.0&tabs=csharp%2Ceml), worked well without magic code)

2) Think it's not possible to put in one single connection "all the necessary scopes that are needed for all the Skills" (as stated [here](https://microsoft.github.io/botframework-solutions/overview/whats-new/1.0/enable-sso-with-skills-using-oauthcredentials-setting/)) if those scopes refers to more than one resources.

For 2) I tried using separate connections for each resource. In that case login works but not SSO: a separate login prompt appears for each skill.

Does exists, or it is possible to have, a more detailed guide on how to have SSO in VA+Skills scenario (more than one skill, each accessing a different resource)?

It will be great to have it. Thanks in advance.

Ciao,
Alessandro.

**Comment by GitDruid** [comment]
Hi Ted @lzc850612 ,
in addition to my previous question, is there a plan for the release of the next SDK version that will include the SSO in the core?

Thanks,
Alessandro.

**Comment by GitDruid** [comment]
@lzc850612, @lauren-mills  any news about this issue/my questions?

**Comment by lzc850612** [comment]
@GitDruid I just checked the latest SDK code and I'm not seeing SSO being added.

@carlosscastro carlos are you aware if there's a plan to add SSO capability into SDK SkillHandler so SDK can natively support SSO without developers having to handle this on their own?

**Comment by GitDruid** [comment]
No news about this?

**Comment by cwhitten** [comment]
@peterinnesmsft is this still being worked on?

**Comment by peterinnesmsft** [comment]
This is not actively being worked on. I'll take a look into this and follow up once I understand where things were left off.

**Comment by tonyanziano** [comment]
@peterinnesmsft can you please provide an update on the progress of this?

**Comment by Batta32** [comment]
Hi @GitDruid , sorry for the delay. We are currently reviewing this issue and we will let you know as soon as we have any update for you 😊.

**Comment by Batta32** [comment]
Hi @GitDruid, sorry for the delay. We reproduced this issue following the Microsoft [documentation](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-authentication-sso?view=azure-bot-service-4.0&tabs=csharp%2Ceml), and we configured the Virtual Assistant and the Skills, along with their corresponding AppRegistration and OAuth settings.

Following your [comment](https://github.com/microsoft/botframework-solutions/issues/3579#issuecomment-663121663), we enabled the tokenExchangeConfig in our Virtual Assistant, but when we tried to call our Skills we received the internalServerError that you attached, even if we use the change of #3667.
Regarding the magic code you mentioned we also have it, we tried to avoid it using Direct Line Enhanced Authentication, as this [document](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-security-enhanced?view=azure-bot-service-4.0) explains, but with no success.

As soon as we have any update we will get back to you here 😊

**Comment by srinaath** [comment]
@Batta32  Any updates on this ticket?

**Comment by Batta32** [comment]
Hi @srinaath, we are not currently working on this issue. The next week we will probably re-take it and as soon as we have any update, we will communicate it here 😊.

**Comment by VictorGrycuk** [comment]
@gabog has the SDK team created an in depth guide regarding the SSO? Specifically speaking about using SSO between a Virtual Assistant and multiple Skills, where each skill has a different set of scopes/resources.

We found [this document](https://github.com/microsoft/botbuilder-dotnet/blob/main/tests/Skills/ReadMeForSSOTesting.md) that looks part of a sample.

**Comment by congysu** [comment]
@gabog could you please take a look at this one?  Thanks.

**Comment by carlosscastro** [comment]
@gabog can we set up some time to assess SSO docs with the documentation team and make a plan to get them right?

**Comment by Kmeredith-hcg** [comment]
Is there any update on this issue? My company is needing this as well.
